src/views/projects/SourceBrowser/FileDiff.svelte

```diff
@@ -58,6 +114,90 @@
       }
     }
   }
+
+  function isLineSelected(
+    selection: Selection | undefined,
+    hunkIdx: number,
+    lineIdx: number,
+  ): boolean {
+    if (!selection) {
+      return false;
+    }
+
+    if (selection.endHunk !== undefined && selection.endLine !== undefined) {
+      return (
+        hunkIdx >= selection.startHunk &&
+        hunkIdx <= selection.endHunk &&
+        (hunkIdx === selection.startHunk
+          ? lineIdx >= selection.startLine
+          : true) &&
+        (hunkIdx === selection.endHunk ? lineIdx <= selection.endLine : true)
+      );
+    } else {
+      return hunkIdx === selection.startHunk && lineIdx === selection.startLine;
+    }
+  }
+
+  function shouldScroll(hunkIdx: string, lineIdx: string): boolean {
+    if (!selection) {
+      return false;
+    }
+
+    return (
+      hunkIdx === selection.startHunk.toString() &&
+      lineIdx === selection.startLine.toString()
+    );
+  }
+
+  function calcNewHash(
+    path: string,
+    hunkIdx: number,
+    lineIdx: number,
+    shift: boolean,
+  ): string {
+    // single line selection
+    if (!shift) {
+      return path + ":H" + hunkIdx + "L" + lineIdx;
+    }
+
+    if (!selection) {
+      return "";
+    }
+
+    // range selection
+    if (hunkIdx === selection.startHunk) {
+      if (lineIdx >= selection.startLine) {
+        return `${path}:H${hunkIdx}L${selection.startLine}H${hunkIdx}L${lineIdx}`;
+      } else {
+        return `${path}:H${hunkIdx}L${lineIdx}H${hunkIdx}L${selection.startLine}`;
+      }
+    } else if (hunkIdx < selection.startHunk) {
+      return `${path}:H${hunkIdx}L${lineIdx}H${selection.startHunk}L${selection.startLine}`;
+    } else {
+      return `${path}:H${selection.startHunk}L${selection.startLine}H${hunkIdx}L${lineIdx}`;
+    }
+  }
+
+  function updateHash(newHash: string) {
+    if (newHash !== "") {
+      window.location.hash = newHash;
+    } else {
+      window.history.replaceState(
```

Comment:
With the implementation of the custom page router, we started to avoid leaking `window.history` into individual components.
Could we maybe move this function to `src/lib/router.ts` or solve this without having `window.history` in this component?

Reply:
I can remove `window.history` if you have a solution for this: https://github.com/radicle-dev/radicle-interface/pull/942#issuecomment-1638536224

Reply:
@xphoniex try the event modifier self: `on:click|self`

Reply:
@hecatia-elegua that's not the only issue, `window.history.replaceState | .pushState` doesn't cause `hashchange` to be emitted, so while the url changes, selection is still visible and `FileDiff` doesn't update.

Reply:
Maybe you can `updateSelection` after `replaceState, or `updateSelection` in the `on:click|self` as well, I'm not really sure. I've tried around a bunch.

Also you probably don't want `<main>`, since it would only trigger at the screen edges, but probably the .metadata `<div>` or some other way to do this

Reply:
works, added in https://github.com/radicle-dev/radicle-interface/pull/942/commits/d1c5916da5581f26fe02e54c67ae17b80fe7ab4b

Reply:
> With the implementation of the custom page router, we started to avoid leaking `window.history` into individual components.

I share your concern, @sebastinez. But I do think the hack is fine for the moment.